### PR TITLE
Use more slice patterns inside the compiler

### PR DESCRIPTION
**Pull request by fmease**
Nothing super noteworthy. Just replacing the common 'fragile' pattern of "length check followed by indexing or unwrap" with slice patterns for legibility and 'robustness'.

r? ghost

### CODE REVIEW

**Comment by fmease** [comment]
r? compiler

File: compiler/rustc_ast/src/ast.rs [hunk 1]
```diff
@@ -585,7 +585,7 @@ impl Pat {
             }
             // A slice/array pattern `[P]` can be reparsed as `[T]`, an unsized array,
             // when `P` can be reparsed as a type `T`.
-            PatKind::Slice(pats) if pats.len() == 1 => pats[0].to_ty().map(TyKind::Slice)?,
+            PatKind::Slice(pats) if let [pat] = &**pats => pat.to_ty().map(TyKind::Slice)?,
```
Comment (compiler-errors) [inline compiler/rustc_ast/src/ast.rs:588]: I would kinda prefer we spell this as `pats.as_slice()` or `&pats[..]`.

`&**` is kinda an eyesore, and the implicit autoderef of either strategies makes it less annoying if we change around the underlying type reprs.

File: compiler/rustc_mir_build/src/thir/cx/expr.rs [hunk 2]
```diff
@@ -292,9 +292,9 @@ impl<'tcx> Cx<'tcx> {
                 } else {
                     let attrs = tcx.hir().attrs(expr.hir_id);
                     if attrs.iter().any(|a| a.name_or_empty() == sym::rustc_box) {
-                        if attrs.len() != 1 {
+                        if let [attr] = attrs {
```
Comment (compiler-errors) [inline compiler/rustc_mir_build/src/thir/cx/expr.rs:295]: You need to invert this check.

File: compiler/rustc_mir_transform/src/validate.rs [hunk 3]
```diff
@@ -898,8 +898,8 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                         self.param_env,
                         adt_def.non_enum_variant().fields[field].ty(self.tcx, args),
                     );
-                    if fields.len() == 1 {
-                        let src_ty = fields.raw[0].ty(self.body, self.tcx);
+                    if let [field] = &*fields.raw {
```
Comment (compiler-errors) [inline compiler/rustc_mir_transform/src/validate.rs:901]: :( this is kinda annoying to have to do, but I guess at least equally annoying as `fields.raw[0]`.

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**